lib/matplotlib/axes/_axes.py

```diff
@@ -475,9 +481,25 @@ def inset_axes(self, bounds, *, transform=None, zorder=5,
 
         return inset_ax
 
-    def indicate_inset(self, bounds, inset_ax=None, *, transform=None,
-            facecolor='none', edgecolor='0.5', alpha=0.5,
-            zorder=4.99, **kwargs):
+    def indicate_inset(
+            self,
+            bounds: Tuple[float, float, float, float],
+            inset_ax: Optional['Axes'] = None,
+            *,
+            transform: Optional[mtransforms.Transform] = None,
+            facecolor='none',
+            edgecolor='0.5',
+            alpha: float = 0.5,
+            zorder: float = 4.99, **kwargs
+        ) -> Tuple[
```

Comment:
I've not ever seen this syntax before.  Are we supporting it in matplotlib?  @anntzer @timhoffm ?

Reply:
These are [PEP-484 type hints](https://www.python.org/dev/peps/pep-0484/); they're immensely helpful to downstream users (in my opinion). Matplotlib which has a deep, nested class structure, and the docs can be tricky to navigate: knowing exactly what you're getting out of a function call, and your IDE being able to tell you if you're using it wrong, increases discoverability of the library and productivity of the user. A systematic, well-specified, core python way of documenting types in signatures are also just better than the somewhat vague numpydoc conventions for docstrings. Sphinx can pick it up with an extension, the whole module path is necessarily resolved, and the IDE support is much broader. We've been messing around with rst, numpydoc, googledoc for years: now we have one obvious way to do it.

It might be nice if eventually most of matplotlib's API surface was covered, but there's no need to do it all in one go - because it's not checked at runtime and not enforced they can be phased in gradually. As this is new API surface which will only ever be used in python 3.6+, it seemed a good corner of the code to start!

Reply:
Given that @jklymak asked me for my opinion... I personally think the design of PEP484 (and followup PEPs, especially 560, 561, 563) type hints is poor and am not particularly looking forward to adding type hints everywhere (no, I don't have any concrete alternative proposal).
On the other hand, I guess type hints are going one day or another to be everywhere :/, so I'm not going to *block* their addition either.  (I guess some here may be a bit surprised to hear that there actually are some new and shiny Python features that I *don't* like :p)

(FWIW I would have been more interested by work on type inference (à la C++ "almost always auto" style).)

Reply:
I think we need an internal decision if and how we want to add type hints. See also @tacaswell's comment: https://github.com/matplotlib/matplotlib/issues/13798#issuecomment-477979607.